27). Add topic files 79 to 82 (#18081)
dotnet/docs: docs/standard/linq/filter-attribute.md

---
title: How to filter on an attribute - LINQ to XML

> description: This article shows how to use LINQ to XML query and XPath, in C# and Visual Basic, to find descendant elements that have a specified name and attribute value.

Reviewer: ```suggestion
description: Learn how to use LINQ to XML query and XPath, in C# and Visual Basic, to find descendant elements that have a specified name and attribute value.
```

ms.date: 07/20/2015
dev_langs:
  - "csharp"
  - "vb"
ms.assetid: 208d6256-1bd7-4237-b2c9-909f26dfd0e2
---

# How to filter on an attribute (LINQ to XML)

This article shows how to use LINQ to XML query and XPath, in C# and Visual Basic, to find descendant elements that have a specified name and attribute value.

## Example: Find all descendant elements that have a specified name and attribute value

> This example uses LINQ to XML query and XPath to find, in XML document [Sample XML file: Multiple purchase orders](sample-xml-file-multiple-purchase-orders.md), all descendant elements that have the name `Address`, and a `Type` attribute whose value is "Shipping". The XPath expression is `.//Address[@Type='Shipping']`

Reviewer: The way that the XML file is referenced here is a bit odd. 

```suggestion
This example uses LINQ to XML query and XPath to find all descendant elements that have the name `Address`, and a `Type` attribute whose value is "Shipping" in an XML document. The XPath expression is `.//Address[@Type='Shipping']`. The XML document used is located at [Sample XML file: Multiple purchase orders](sample-xml-file-multiple-purchase-orders.md).
```

```csharp
XDocument po = XDocument.Load("PurchaseOrders.xml");

// LINQ to XML query
IEnumerable<XElement> list1 =
    from el in po.Descendants("Address")
    where (string)el.Attribute("Type") == "Shipping"
    select el;

// XPath expression
IEnumerable<XElement> list2 = po.XPathSelectElements(".//Address[@Type='Shipping']");

if (list1.Count() == list2.Count() &&
        list1.Intersect(list2).Count() == list1.Count())
    Console.WriteLine("Results are identical");
else
    Console.WriteLine("Results differ");
foreach (XElement el in list1)
    Console.WriteLine(el);
```

```vb
Dim po As XDocument = XDocument.Load("PurchaseOrders.xml")

' LINQ to XML query
Dim list1 As IEnumerable(Of XElement) = _
    From el In po...<Address> _
    Where el.@Type = "Shipping" _
    Select el

' XPath expression
Dim list2 As IEnumerable(Of XElement) = _
    po.XPathSelectElements(".//Address[@Type='Shipping']")

If (list1.Count = list2.Count And _
        list1.Intersect(list2).Count() = list1.Count()) Then
    Console.WriteLine("Results are identical")
Else
    Console.WriteLine("Results differ")
End If
For Each el As XElement In list1
    Console.WriteLine(el)
Next
```

> The example produces this output:

Reviewer: ```suggestion
The example produces the following output:
```
